packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+      const method = request.method || 'GET';
+
+      plugin._logger.debug('%s plugin incomingRequest', plugin.moduleName);
+
+      if (
+        Utils.isIgnored(pathname, request, plugin.options.ignoreIncomingPaths)
+      ) {
+        return original.apply(this, [event, ...args]);
+      }
+
+      const propagation = plugin._tracer.getHttpTextFormat();
+      const headers = request.headers;
+      const spanOptions: SpanOptions = {
+        kind: SpanKind.SERVER,
+      };
+
+      const spanContext = propagation.extract(Format.HTTP, headers);
+      if (spanContext && isValid(spanContext)) {
+        spanOptions.parent = spanContext;
+      }
+
+      const span = plugin._startHttpSpan(`${method} ${pathname}`, spanOptions);
+
+      return plugin._tracer.withSpan(span, () => {
+        plugin._tracer.bind(request);
+        plugin._tracer.bind(response);
+
+        // Wraps end (inspired by:
+        // https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/src/plugins/plugin-connect.ts#L75)
+        const originalEnd = response.end;
+        response.end = function(
+          this: ServerResponse,
+          ...args: ResponseEndArgs
+        ) {
+          response.end = originalEnd;
+          // Cannot pass args of type ResponseEndArgs,
+          // tslint complains "Expected 1-2 arguments, but got 1 or more.", it does not make sense to me
+          // tslint:disable-next-line:no-any
+          const returned = response.end.apply(this, arguments as any);
+          const requestUrl = request.url ? url.parse(request.url) : null;
+          const hostname = headers.host
+            ? headers.host.replace(/^(.*)(\:[0-9]{1,5})/, '$1')
+            : 'localhost';
+          const userAgent = (headers['user-agent'] ||
+            headers['User-Agent']) as string;
+
+          const attributes: Attributes = {
+            [AttributeNames.HTTP_URL]: Utils.getUrlFromIncomingRequest(
+              requestUrl,
+              headers
+            ),
+            [AttributeNames.HTTP_HOSTNAME]: hostname,
+            [AttributeNames.HTTP_METHOD]: method,
+            [AttributeNames.HTTP_STATUS_CODE]: response.statusCode,
+            [AttributeNames.HTTP_STATUS_TEXT]: response.statusMessage,
+          };
+
+          if (requestUrl) {
+            attributes[AttributeNames.HTTP_PATH] = requestUrl.path || '/';
+            attributes[AttributeNames.HTTP_ROUTE] = requestUrl.pathname || '/';
+          }
+
+          if (userAgent) {
+            attributes[AttributeNames.HTTP_USER_AGENT] = userAgent;
+          }
+
+          span
+            .setAttributes(attributes)
+            .setStatus(Utils.parseResponseStatus(response.statusCode));
+
+          if (plugin.options.applyCustomAttributesOnSpan) {
+            plugin.options.applyCustomAttributesOnSpan(span, request, response);
+          }
+
+          span.end();
+          return returned;
+        };
+        return original.apply(this, [event, ...args]);
```

Comment:
if `original` throws the span is never ended.

Reply:
The `original` in this case is the `emit` function which calls the user `request` handler. I agree that this handler should usually not throw but it's out of control of node.js and OT.
`BasicTracer.withSpan()` just delegates to it's `ScopeManager.with()` which has no `try/catch`.

dump sample: `http.createServer((req, res) => { throw new Error("Boom"); }).listen(8000)`

Reply:
Yes, I think a `try/catch` block and rethrow the exception in the catch block after cleanup OT specifics should be done (or `try/finally` if the error thrown is not needed to end the span).

Created https://github.com/open-telemetry/opentelemetry-js/issues/222 for a general discussion regarding this.

Reply:
I answered in the isue but `AsyncHooksScopeManager` do have try/catch when calling original methods.